Attribute: Implement get<>() Cast

ax3l opened the pull request:
Close #335: Implement `getCast<>()` in `Attribute::get<U>()`. Will now cast
during `::get<U>()` if convertible.

Implementations can now simply use `::get<>()` again :)
Final clean-up for refactoring started in #337

ax3l committed:
Attribute: Implement get<>() Cast

Implement `getCast<>()` in `Attribute::get<U>()`. Will now cast
during get if convertible.

ax3l committed:
Implementations: use Attribute::get<>

getCast is now convieniently implemented in ::get

ax3l committed:
Attribute: Remove return after throw

Added this for being pedantic in case this causes a warning.
Nevertheless, would have to create a factory here to properly
init a `std::array`.

ax3l commented:
@C0nsultant Ping :)

anokfireball commented:
This make [`Attributable::readFloatingPoint()`](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/backend/Attributable.hpp#L244-L275) redundant. The casting logic applied therein is now handled in `get` itself. I'll do a follow-up PR that removes this redundancy and just uses `get`.

anokfireball commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -192,8 +209,12 @@ getCast( Attribute const a )
+template< typename U >
+U Attribute::get() const
+{
+    return getCast< U >( Variant::getResource() );
```
I would hate to cast silently here and not notify the user about possible loss of precision.
There already is a function that does this ([`warnWrongDtype()`](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/src/backend/Attributable.cpp#L282-L290)). I'll push a patch shortly that will use that function if the resource dtype and `U` differ.

ax3l replied:
That's actually not what I would like as a user.
If there is an attribute "attr" I can RT check it's dtype. If I want to `get<>` it as something, I want to get it as exactly such.

ax3l replied:
Let's have a quick chat on the philosophy of that method if you like :) Gitter? VC?
In case you are busy today, we can also chat on Monday. I have to leave here around noon.

anokfireball replied:
Offline resolution:
C++ requires a certain effort from the user to put the *type* inside the brackets. Thus, he should be well aware of what he is requesting.
Information about the datatype can be obtained at run-time. If non-casting behavior is desired, we shift this obligation to the user.

ax3l replied:
> If non-casting behavior is desired, we shift this obligation to the user.

e.g. by checking the `dtype` at runtime, taking the `resource` directly or casting back to the underlying `auxiliary::Variant` that will throw.

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1526,45 +1526,45 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
-    REQUIRE(getCast< std::vector< long double > >(s.getAttribute("vecLongdouble")) == std::vector< long double >({0.L, std::numeric_limits<long double>::max()}));
+    REQUIRE(s.getAttribute("vecLongdouble").get< std::vector< long double > >() == std::vector< long double >({0.L, std::numeric_limits<long double>::max()}));
     REQUIRE(s.getAttribute("vecString").get< std::vector< std::string > >() == std::vector< std::string >({"vector", "of", "strings"}));
     // REQUIRE(s.getAttribute("bool").get< bool >() == true);
```
This should work as expected now, as the stored `uint8_t` can be cast to `bool` in the new `get`.

ax3l replied:
Just need to rebase, done already in #344

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1527,25 +1527,25 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
+    REQUIRE(s.getAttribute("vecLongdouble").get< std::vector< long double > >() == std::vector< long double >({0.L, std::numeric_limits<long double>::max()}));
     REQUIRE(s.getAttribute("vecString").get< std::vector< std::string > >() == std::vector< std::string >({"vector", "of", "strings"}));
     REQUIRE(s.getAttribute("bool").get< unsigned char >() == true);
     REQUIRE(s.getAttribute("boolF").get< unsigned char >() == false);
```
This now can be retrieved as `bool`, thanks to the casting `get`.

Have this in my staged set, will push soon.

ax3l replied:
That sounds great, thanks! :)

anokfireball committed:
Use new casting get to read fp Attributes

The handling an casting of arbitrary-width floating-point Attributes is
now redundant with the casting that is present in Attrbute::get.

ax3l commented on src/Datatype.cpp:
```diff
@@ -22,6 +22,19 @@
+                    Datatype store,
+                    Datatype request)
+{
+    std::cerr << "Warning: Attribute '" << key
```
loving MSVC:
```
error C2679: binary '<<': no operator found which takes a right-hand operand of type 'const std::string' (or there is no acceptable conversion
```

forgot to include `<string>`

ax3l replied:
fixed it :)
